File: cpp/sdk-basics/keys/src/Keys.cpp
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <stdio.h>
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>
#include <map>
using namespace std;

#ifdef _WIN32
    #define HOMEVAR "USERPROFILE"
#else 
    #define HOMEVAR "HOME"
#endif

void print_vector(vector <string> const &strings) {
    bool first_time = true;

    cout << "[";
    for (auto const& str: strings) {
        if (first_time) {
            cout << "\"" << str << "\"";
            first_time = false;
        }
        else {
            cout << ",\"" << str << "\"";
        }
    }
    cout << "]";
}

void print_attributes(map<string, vector<string>> const &attributes) {
    bool first_time = true;

    cout << "{";
    for (auto const& pair: attributes) {
        if (first_time) {
            cout << "\"" << pair.first << "\": ";
            first_time = false;
        }
        else {
            cout << ", \"" << pair.first << "\": ";
        }
        print_vector(pair.second);
    }
    cout << "}";
}

void print_fixed_attrs(map<string, vector<string>> const &fixed_attrs) {
    cout << "FixedAttrs   : ";
    print_attributes(fixed_attrs);
    cout << endl;
}

void print_mutable_attrs(map<string, vector<string>> const &mutable_attrs) {
    cout << "MutableAttrs : ";
    print_attributes(mutable_attrs);
    cout << endl;
}

int main(int argc, char* argv[]) {

    int nErrorCode;

    // read persistor password from environment variable
    char* cpersistorPassword = getenv("IONIC_PERSISTOR_PASSWORD");
    if (cpersistorPassword == NULL) {
        cerr << "[!] Please provide the persistor password as env variable: IONIC_PERSISTOR_PASSWORD" << endl;
        exit(1);
    }
    string persistorPassword = string(cpersistorPassword);

    // initialize agent with password persistor
    string persistorPath = string(getenv(HOMEVAR)) + "/.ionicsecurity/profiles.pw";
    ISAgentDeviceProfilePersistorPassword persistor;
    persistor.setFilePath(persistorPath);
    persistor.setPassword(persistorPassword);
    ISAgent agent;
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        cerr << "Failed to initialize agent from password persistor (" << persistorPath << ")" << endl;
        cerr << ISAgentSDKError::getErrorCodeString(nErrorCode) << endl;
        exit(1);
    }
    agent.setMetadata("ionic-application-name", "ionic-keys-tutorial");
    agent.setMetadata("ionic-application-version", "1.0.0");

    // define fixed attributes
    map<string, vector<string >> fixedAttributes;
    vector<string> dataTypeVal;
    vector<string> regionVal;
    dataTypeVal.push_back("Finance");
    regionVal.push_back("North America");
    fixedAttributes["data-type"] = dataTypeVal;
    fixedAttributes["region"] = regionVal;

    // define mutable attributes
    map<string, vector<string >> mutableAttributes;
    vector<string> classificationVal;
    vector<string> designatedOwnerVal;
    classificationVal.push_back("Restricted");
    designatedOwnerVal.push_back("joe@hq.example.com");
    mutableAttributes["classification"] = classificationVal;
    mutableAttributes["designed_owner"] = designatedOwnerVal;

    // create new key with fixed and mutable attributes
    ISAgentCreateKeysRequest request;
    ISAgentCreateKeysRequest::Key requestKey("refid1", 1, fixedAttributes, mutableAttributes);
    request.getKeys().push_back(requestKey);
    ISAgentCreateKeysResponse response;
    nErrorCode = agent.createKeys(request, response);
    if (nErrorCode != ISAGENT_OK) {
        cerr << "Error creating key: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << endl;
        exit(1);
    }
    const ISAgentCreateKeysResponse::Key *createdKey = response.findKey("refid1");

    // display new key
    cout << "NEW KEY:" << endl;
    ISCryptoHexString hexKeyCreated;
    hexKeyCreated.fromBytes(createdKey->getKey());
    cout << "KeyId        : " << createdKey->getId() << endl;
    cout << "KeyBytes     : " << hexKeyCreated << endl;
    print_fixed_attrs(createdKey->getAttributes());
    print_mutable_attrs(createdKey->getMutableAttributes());
    

    // get key by KeyId
    ISAgentGetKeysResponse getResponse;
    string keyId = createdKey->getId();
    nErrorCode = agent.getKey(keyId, getResponse);
    if (nErrorCode != ISAGENT_OK) {
        cerr << "Error fetching key: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << endl;
        exit(1);
    }
    if (response.getKeys().size() == 0) {
        cerr << "No key was returned (key does not exist or access was denied)" << endl;
        exit(1);
    }
    ISAgentGetKeysResponse::Key fetchedKey = getResponse.getKeys().at(0);

    // display fetched key
    cout << "\nFETCHED KEY:" << endl;
    ISCryptoHexString hexKeyFetched;
    hexKeyFetched.fromBytes(fetchedKey.getKey());
    cout << "KeyId    : " << fetchedKey.getId() << endl;
    cout << "KeyBytes : " << hexKeyFetched << endl;
    print_fixed_attrs(fetchedKey.getAttributes());
    print_mutable_attrs(fetchedKey.getMutableAttributes());

    // define new mutable attributes
    vector<string> newClassificationVal;
    newClassificationVal.push_back("Highly Restricted");

    // merge new and existing mutable attributes
    ISAgentUpdateKeysRequest::Key updateKey(fetchedKey);
    updateKey.getMutableAttributes()["classification"] = newClassificationVal;

    // update key
    ISAgentUpdateKeysResponse updateResponse;
    nErrorCode = agent.updateKey(updateKey, updateResponse);
    if (nErrorCode != ISAGENT_OK) {
        cerr << "Error updating key: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << endl;
        exit(1);
    }
    ISAgentUpdateKeysResponse::Key updatedKey = updateResponse.getKeys().at(0);

    // display updated key
    cout << "\nUPDATED KEY:" << endl;
    ISCryptoHexString hexKeyUpdated;
    hexKeyUpdated.fromBytes(updatedKey.getKey());
    cout << "KeyId    : " << updatedKey.getId() << endl;
    cout << "KeyBytes : " << hexKeyUpdated << endl;
    print_fixed_attrs(updatedKey.getAttributes());
    print_mutable_attrs(updatedKey.getMutableAttributes());
}

File: javascript/sdk-basics/keys/keys.js
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

/*
 ** PRE-REQUISITES **
 * 1. A Machina tenant. You can obtain one at: https://ionic.com/start-for-free/.
 * 2. Your device, in this case the browser, needs to be enrolled. This is done 
 *    when you enrolled your device after signing up for a tenant. Enrollment 
 *    can also be accomplished by executing the Enroll Device script at:
 *    ../../enroll-device/index.html.
 */

import {getAgentConfig} from '../../jssdkConfig.js';

function printKey( text, key ) {
  console.log(text);
  console.log(`KeyId    : ${key.keyId}`);
  console.log(`KeyBytes : ${key.key}`);
  console.log(`FixedAttributes   : ${JSON.stringify(key.attributes, null, 0)}`);
  console.log(`MutableAttributes : ${JSON.stringify(key.mutableAttributes, null, 0)}`);
  console.log(' ');
}

const main = async () => {

  // Get the tutorial application data. This assures all tutorials use the same
  // app ID, user ID and user authentication. It matches what was used for enrollment.
  const appData = getAgentConfig('JavaScript Keys Tutorial');
  let response;

  // Initialize the Machina agent.
  try {
    response = await new window.IonicSdk.ISAgent(appData);
  } catch (errorResp) {
    console.error('Error initializing ionic agent:' + errorResp);
    return;
  }
  const agent = response.agent;

  // Define fixed attributes.
    const fixedAttributes = {
      'data-type': ['Finance'],
      'region': ['North America'],
    };

  // Define mutable attributes.
  let mutableAttributes = {
    'classification': ['Restricted'],
    'designated_owner': ['joe@hq.example.com'],
  };

  // Create single key with fixed and mutable attributes.
  try {
    response = await agent.createKeys({
      quantity: 1,
      attributes: fixedAttributes,
      mutableAttributes: mutableAttributes
    });
  } catch (errorResp) {
    console.log('Error creating key: ' + errorResp);
    return;
  }

  // Save the key ID.
  const createdKey = response.keys[0];
  const keyId = createdKey.keyId;

  // Display new key.
  console.log('');
  printKey('New Key:', createdKey);

  // Fetch key by Key ID.
  try {
    response = await agent.getKeys({keyIds: [keyId]});
  } catch (errorResp) {
      console.log('Error getting key: ' + errorResp);
      return;
  }

  const fetchedKey = response.keys[0];

  // Display fetched key.
  printKey('Fetched Key:', fetchedKey);

  // Define new mutable attributes.
  const newMutableAttributes = {
    'classification': ['Highly Restricted'],
  };

  // Merge new and existing mutable attributes.
  Object.assign(mutableAttributes, newMutableAttributes);

  // Update the key.
  try {
    response = await agent.updateKeys({
      keyRequests: [{
        keyId: fetchedKey.keyId,
        force: true,
        mutableAttributes: mutableAttributes,
      }],
    });
    } catch (errorResp) {
      console.log('Error updating key: ' + errorResp);
      return;
  }

  const updatedKeyId = response.keys[0].keyId;

  // To display the updated key, a fetch key is required.
  try {
    response = await agent.getKeys({
      keyIds: [updatedKeyId],
      });
  } catch (errorResp) {
    console.log('Error fetching updated key: ' + errorResp);
    return;
  }

  const updatedKey = response.keys[0];

  // Display updated key.
  printKey('Updated Key:', updatedKey);

};

main();
